src/axom/inlet/Table.hpp

```diff
@@ -19,17 +19,225 @@
+ * from inlet with axom::inlet::Table::get(const std::string&)
+ * 
+ * This is the only way of reading in a non-default-constructible type from
+ * Inlet.
+ * 
+ * \see axom::inlet::Table::get(const std::string&)
+ *******************************************************************************
+ */
+template <typename T>
+struct FromInlet
+{
+  // "Poison" the base implementation so it cannot be used
+  // and so it can be checked in a SFINAE context
+  FromInlet() = delete;
+};
+
 namespace axom
 {
 namespace inlet
 {
+namespace detail
+{
+/*!
+ *******************************************************************************
+ * \class is_inlet_primitive
+ *
+ * \brief A type trait for checking if a type is isomorphic to a Lua primitive
+ * \tparam T The type to check
+ *******************************************************************************
+ */
+template <typename T>
+struct is_inlet_primitive
+{
+  using BaseType = typename std::decay<T>::type;
+  static constexpr bool value = std::is_same<BaseType, bool>::value ||
+    std::is_same<BaseType, int>::value || std::is_same<BaseType, double>::value ||
+    std::is_same<BaseType, std::string>::value;
+};
+
+/*!
+ *******************************************************************************
+ * \class is_inlet_primitive
+ *
+ * \brief A type trait for checking if a type is isomorphic to an array of Lua
+ * primitives
+ * \tparam T The type to check
+ *******************************************************************************
+ */
+template <typename T>
+struct is_inlet_primitive_array : std::false_type
+{ };
+
+// If it's an unordered map whose value type is a lua primitive,
+// assume that it's an array
+template <typename T>
+struct is_inlet_primitive_array<std::unordered_map<int, T>>
+{
+  static constexpr bool value = is_inlet_primitive<T>::value;
+};
+
+/*!
+ *******************************************************************************
+ * \class has_FromInlet_specialization
+ *
+ * \brief A type trait for checking if a type has specialized FromInlet
+ * with the required T operator()(axom::inlet::Table&)
+ * \tparam T The type to check
+ *******************************************************************************
+ */
+template <typename T, typename SFINAE = void>
+struct has_FromInlet_specialization : std::false_type
+{ };
+
+template <typename T>
+struct has_FromInlet_specialization<
+  T,
+  typename std::enable_if<
+    std::is_same<T, decltype(std::declval<FromInlet<T>&>()(std::declval<Table&>()))>::value>::type>
+  : std::true_type
+{ };
+
+}  // namespace detail
+
+/*!
+ *******************************************************************************
+ * \class Proxy
+ *
+ * \brief Provides a uniform interface for access and conversion to primitive
+ * and user-defined types
+ *
+ * \see Inlet Field
+ * \see Inlet Table
+ *******************************************************************************
+ */
+class Proxy
```

Comment:
I like to keep one class per file unless there is a reason.  Can you move this into its own file?